lib/Driver/Compilation.cpp

```diff
@@ -513,8 +533,18 @@ namespace driver {
           // Parseable output was requested.
           FinishedCmd->forEachContainedJobAndPID(Pid, [&](const Job *J,
                                                           Job::PID P) {
-            parseable_output::emitFinishedMessage(llvm::errs(), *J, P,
-                                                  ReturnCode, Output, ProcInfo);
+            if (ReturnCode != 0 &&
+                isBatchJob(FinishedCmd) &&
+                batchConstituentJobProducedZeroLengthSerializedDiagnostics(J)) {
```

Comment:
The three conjuncts in this `if` deserve to be factored out. There is an intension-crossing here:  as I read it, the `if` is really testing something like `wasJobIncomplete`, so I would factor out the predicate into a (static?) function testing the three values and returning a bool. Then there could be a nice Doxygen comment for that function explaining a bit.
I would also be tempted (in addition) to factor out the whole paragraph from lines 536 to 547 into an parseable_output static function called something like `emitAppropriateMessage` in order to reflect the intentionality here, as well as to keep the enclosing context from growing any larger.

Reply:
fixed

Reply:
Great!